136287611 - Allowed file types

nisevi opened the pull request:
PT Story: https://www.pivotaltracker.com/story/show/136287611

Changes proposed in this pull request:
1. Mime types allowed on 'UploadedFile' model are picked from constant;
2. Same constant used on controller to let the view know which file extensions are allowed;

Screenshots:
![image](https://user-images.githubusercontent.com/3024789/35779336-aad7fdca-09cb-11e8-8546-0a2512280b54.png)

nisevi committed:
Added 'upload_allowed_file_types' to locale files for English and Swedish.

nisevi committed:
Listed allowed file types from constant created to hold mime types and extensions on 'UploadedFile' model.

nisevi committed:
Setup allowed file types, only where needed.

patmbolger commented on app/views/shf_applications/_shf_application_fields.html.haml:
```diff
@@ -46,11 +46,15 @@
 
       - submit_label = @shf_application.new_record? ? t('submit') : t('save')
+      %p= t('shf_applications.new.upload_allowed_file_types')
+      %p= @allowed_file_types
```
Please add an `accept` option to the `file_field` (line 51) so that the user can only select valid file types in the upload-file dialog box.

If you find a nice way to test this added item, it would be great.  But, that is not critical for this.

patmbolger replied:
BTW, adding the above item would also address https://www.pivotaltracker.com/story/show/153360586.

nisevi replied:
@patmbolger the `accept` option was added to the `file_field` for the user to only see the allowed file types.

nisevi committed:
Only display to the user the allowed file types when trying to upload a file.

nisevi commented:
Allowed file type extensions are being displayed to the user:

![image](https://user-images.githubusercontent.com/3024789/35848603-009e6f1a-0b1f-11e8-9400-8879fc6fdb2b.png)

Only display to the user the allowed files that can be uploaded:

![image](https://user-images.githubusercontent.com/3024789/35848739-6f995ee8-0b1f-11e8-9596-87df8e2919f2.png)

This are all the files on that folder:

![image](https://user-images.githubusercontent.com/3024789/35848757-8460e12a-0b1f-11e8-8557-0efd269044bd.png)

nisevi commented:
`Cucumber` is failing on `SemaphoreCI` because I think that is trying to find a type of file that we are not displaying any more, since we are specifying which file types the user should see as default. I mean the user always can select `All files` and then it will see other file types to upload, so I'm working on the cucumber tests now :+1:

nisevi commented:
These are the failing scenarios:

```
Failing Scenarios:
cucumber features/create_membership_application.feature:133 # Scenario Outline: Apply for membership - when things go wrong, Scenarios (#1)
cucumber features/create_membership_application.feature:134 # Scenario Outline: Apply for membership - when things go wrong, Scenarios (#2)
cucumber features/create_membership_application.feature:135 # Scenario Outline: Apply for membership - when things go wrong, Scenarios (#3)
cucumber features/create_membership_application.feature:136 # Scenario Outline: Apply for membership - when things go wrong, Scenarios (#4)
cucumber features/create_membership_application.feature:137 # Scenario Outline: Apply for membership - when things go wrong, Scenarios (#5)
cucumber features/create_membership_application.feature:138 # Scenario Outline: Apply for membership - when things go wrong, Scenarios (#6)
cucumber features/create_membership_application.feature:153 # Scenario Outline: Apply for membership: company number wrong length, Scenarios (#1)
cucumber features/create_membership_application.feature:156 # Scenario: Cannot change locale if there are errors in the new application
cucumber features/edit_shf_application.feature:35 # Scenario: Applicant makes mistake when editing his own application
cucumber features/edit_shf_application.feature:90 # Scenario: Cannot change locale if there are errors in the application
cucumber features/emails/new-application-email-to-admins.feature:74 # Scenario: User submits a new application app with bad info so it is not created, so no email sent [SAD PATH]
cucumber features/emails/new-application-gets-ack-email.feature:38 # Scenario: User submits a new application app with bad info so it is not created, so no email sent [SAD PATH]
cucumber features/size_validation.feature:18 # Scenario: New application - Uploads a file that is too large
cucumber features/size_validation.feature:42 # Scenario: Existing application - Uploads a file that is too large
cucumber features/size_validation.feature:60 # Scenario: Size error message in English when locale = en
cucumber features/size_validation.feature:71 # Scenario: Size error message in sveriges when locale = sv
cucumber features/size_validation.feature:80 # Scenario: Switching locales - Size error message in the right language
cucumber features/upload.feature:92 # Scenario: Try to upload a file with unacceptable content type
cucumber features/view_all_companies.feature:143 # Scenario: Visitor sees multiple regions and kommuns for a company in the companies list
cucumber features/view_all_companies.feature:209 # Scenario: Pagination: Set number of items per page

```

nisevi committed:
I added 'update' and 'create' actions to the before action 'set_allowed_file_types'.

The reason for adding 'update' action it happens when the user
tries to upload a file that doesn't match the criteria we have to
accept such file type, then when back on the view our instance object was nil.

Same applies for the 'create' action, and the file being the wrong type.

nisevi commented:
The problem wasn't related to my previous assumption, the real problem was that on `update` and `create` actions, the object holding the allowed types didn't exist, and when the file type was wrong the view was trying to use it.

nisevi committed:
Don't let invalid object to be displayed as 'uploaded' after failed upload.

nisevi commented on app/controllers/shf_applications_controller.rb:
```diff
@@ -218,6 +218,7 @@ def new_file_uploaded(params)
                                            filename: @uploaded_file.actual_file_file_name))
           successful = successful & true
         else
+          @shf_application.uploaded_files.delete(@uploaded_file)
```
@patmbolger I deleted the object because even if the `@uploaded_file` was invalid (here we are in the `else` statement for the `@uploaded_file.valid?` validation) it was hanging there as part of the uploaded files when we were back at the view. Please let me know if you consider that this change should be done in another PR.

patmbolger replied:
No problem - that's fine doing it here.

patmbolger commented:
These changes look good.

In addition, please use the list of allowed file types (from the model) in the rspec validation test.

Right now these are repeated in the test:
```
it { should validate_attachment_content_type(:actual_file)
                    .allowing('image/jpeg', 'image/gif', 'image/png',
                              'text/plain',
                              'text/rtf',
                              'application/pdf',
                              'application/msword',
                              'application/vnd.openxmlformats-officedocument.wordprocessingml.document',
                              'application/vnd.ms-word.document.macroEnabled.12')
                    .rejecting('bin', 'exe') }
```

patmbolger reviewed (changes requested):
Please see comment re change to the rspec test.

nisevi committed:
Using the list of allowed file types (from the model) in the rspec validation test.

nisevi commented:
@patmbolger I updated the rspec test for using the use the list of allowed file types from the model :+1:

nisevi committed:
Merge branch 'develop' into 136287611_allowed_file_types
